### ISSUE TITLE
Fitting band edge effective mass

### ISSUE THREAD

**Issue by 111222111777888** [111222111777888/mTBs#1]
# plot bands
#https://eng.libretexts.org/Bookshelves/Electrical_Engineering/Electronics/Introduction_to_Nanoelectronics_(Baldo)/04%3A_Two_Terminal_Quantum_Wire_Devices/4.12%3A_Effective_Mass

import numpy as np
import matplotlib.pyplot as plt
from sklearn.metrics import r2_score
from scipy.optimize import curve_fit

import os

from matplotlib import rc

rc("font", **{"family": "sans-serif", "sans-serif": ["Helvetica"]})
rc("text")
plt.rcParams["lines.linewidth"] = 2

plt.rcParams["font.size"] = 25
plt.rcParams["legend.fontsize"] = 25
plt.rcParams["figure.figsize"] = (8, 6)

#### get bands
def get_Bands(filename, E_f):
        Data = np.loadtxt(filename)

        X = Data[:,0]
        Y = Data[:,1]

        no_Bands = int(len(X)/len(np.unique(X)))
        print ("No bands is:", no_Bands)

        X_len = len(np.unique(X))
        Bands = []
        Cur_Band = []

        for i in range (0, len(Y)):
                Y[i] = Y[i] - E_f

        for i in range (0, no_Bands):
                Cur_Band = []
                for j in range(0, X_len):
                        Cur_Band.append(Y[ i*X_len + j] )
                Bands.append(Cur_Band)

        X = np.unique(X)

        return (X, Bands, no_Bands)

a_0 = 3.16
X, Bands, no_Bands = get_Bands("Bands.dat.gnu", 11.6250)

# Points
Pts = np.array([0.0000, 0.2545/2, 0.2545] )
Labels = ["$\Gamma$", "A", "$\Gamma$"]

conv = 2 * np.pi / a_0
X = conv * X

ev_to_J = 1.6022e-19
A_to_m = 1e10
Band = np.array(Bands[50])

# K
kpt = 0
krange = 6

X = A_to_m * X
Band = ev_to_J * Band

X = X[kpt:krange] - X[kpt]
Band = Band[kpt:krange]
plt.plot(X, Band, '-', color = 'blue')
#plt.show()

m_e = 9.1093837e-31
hbar = 1.054571817e-34

def quad_fit (k, E_0, m):
        pre = hbar**2/(2*m_e)
        C = pre/m

        E = E_0 + C*k*k

        return (E)

E_0 =  -1.0807000000000002
E_0 = ev_to_J * E_0
m = -0.53

plt.plot(X, quad_fit(X, E_0, m), '-', color = 'orange')
plt.show()

print ("r2: ", r2_score(quad_fit(X, E_0, m), Band))
print ("effective mass:", m)
